src/cockpit/channels/filesystem.py

```diff
@@ -268,3 +287,226 @@ def do_close(self):
+
+        return get_attrs
+
+    def send_update(self, updates: JsonDict, *, reset: bool = False) -> None:
+        if reset:
+            if set(self.current_value) & set(updates):
+                # if we have an overlap, we need to do a proper reset
+                self.send_json({name: None for name in self.current_value}, partial=True)
+                self.current_value = {'partial': True}
+                updates.update(partial=None)
+            else:
+                # otherwise there's no overlap: we can just remove the old keys
+                updates.update({key: None for key in self.current_value})
+
+        json_merge_and_filter_patch(self.current_value, updates)
+        if updates:
+            self.send_json(updates)
+
+    def process_update(self, updates: 'set[str]', *, reset: bool = False) -> None:
+        assert self.fd is not None
+
+        entries: JsonDict = {name: self.getattrs(self.fd, name) for name in updates}
+        info = entries.pop('', {})
+        assert isinstance(info, dict)  # fstat() will never fail with FileNotFoundError
+
+        if self.effective_fnmatch:
+            info['entries'] = entries
+
+        self.send_update({'info': info}, reset=reset)
+
+    def process_pending_updates(self) -> None:
+        assert self.pending is not None
+        if self.pending:
+            self.process_update(self.pending)
+        self.pending = None
+
+    def interesting(self, name: str) -> bool:
+        if name == '':
+            return True
+        else:
+            # only report updates on entry filenames if we match them
+            return fnmatch.fnmatch(name, self.effective_fnmatch)
+
+    def schedule_update(self, name: str) -> None:
+        if not self.interesting(name):
+            return
+
+        if self.pending is None:
+            asyncio.get_running_loop().call_later(0.1, self.process_pending_updates)
+            self.pending = set()
+
+        self.pending.add(name)
+
+    def report_error(self, err: int) -> None:
+        if err == errno.ENOENT:
+            problem = 'not-found'
+        elif err in (errno.EPERM, errno.EACCES):
+            problem = 'access-denied'
+        elif err == errno.ENOTDIR:
+            problem = 'not-directory'
+        else:
+            problem = 'internal-error'
+
+        self.send_update({'error': {
+            'problem': problem, 'message': os.strerror(err), 'errno': errno.errorcode[err]
+        }}, reset=True)
+
+    def flag_onlydir_error(self, fd: Handle) -> bool:
+        # If our requested path ended with '/' then make sure we got a
+        # directory, or else it's an error.  open() will have already flagged
+        # that for us, but systemd_ctypes doesn't do that (yet).
+        if not self.watch or not self.path.endswith('/'):
+            return False
+
+        buf = os.fstat(fd)  # this should never fail
+        if stat.S_IFMT(buf.st_mode) != stat.S_IFDIR:
+            self.report_error(errno.ENOTDIR)
+            return True
```

Comment:
From a purely logical standpoint: what you say is true, but only with `follow: false`.  Otherwise, we'll have resolved the path to a directory already in `open()` and we'll have statted the directory here.  If `follow: false` *is* specified and the `path` is a symlink then of course `/` into that symlink will fail, and that's the desired result.

From a more pragmatic standpoint: this code path only hits on `watch` mode (because of the limitation in systemd_ctypes mentioned in the comment above).  `watch` mode is currently incompatible with `follow: false`, so we're never going to have an fd pointing to a symlink here.